playwright/src/main/java/com/microsoft/playwright/junit/impl/BrowserExtension.java

```diff
@@ -0,0 +1,81 @@
+package com.microsoft.playwright.junit.impl;
+
+import com.microsoft.playwright.Browser;
+import com.microsoft.playwright.BrowserType;
+import com.microsoft.playwright.Playwright;
+import com.microsoft.playwright.PlaywrightException;
+import com.microsoft.playwright.junit.Options;
+import org.junit.jupiter.api.extension.*;
+
+import static com.microsoft.playwright.junit.impl.ExtensionUtils.hasUsePlaywrightAnnotation;
+
+public class BrowserExtension implements ParameterResolver, AfterAllCallback {
+  private static final ThreadLocal<Browser> threadLocalBrowser = new ThreadLocal<>();
+
+  @Override
+  public void afterAll(ExtensionContext extensionContext) {
+    Browser browser = threadLocalBrowser.get();
+    threadLocalBrowser.remove();
+    if (browser != null) {
+      browser.close();
+    }
+  }
+
+  @Override
+  public boolean supportsParameter(ParameterContext parameterContext, ExtensionContext extensionContext) throws ParameterResolutionException {
+    if (!hasUsePlaywrightAnnotation(extensionContext)) {
+      return false;
+    }
+    Class<?> clazz = parameterContext.getParameter().getType();
+    return Browser.class.equals(clazz);
+  }
+
+  @Override
+  public Object resolveParameter(ParameterContext parameterContext, ExtensionContext extensionContext) throws ParameterResolutionException {
+    return getOrCreateBrowser(extensionContext);
+  }
+
+  static Browser getOrCreateBrowser(ExtensionContext extensionContext) {
+    Browser browser = threadLocalBrowser.get();
```

Comment:
This is good as the first step. `Browser` and `Playwright` instances could be reused by tests that not run concurrently (to avoid unsafe multithreaded access) if we want. We do this in .net port already. Not necessary for this PR, just a thought for the future iterations.

Reply:
yea let's think about this.  this will require us keep track of the currently used playwrights and the threads that they're running on (I think).  something for the future.